### PR TITLE
[Core] Optimize distance calculation in `GeometricalObjectBins`

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Reopens https://github.com/KratosMultiphysics/Kratos/pull/12588 after merge conflict

This pull request introduces an optimization to the distance calculation algorithm in the `GeometricalObjectBins` class. The new approach reduces redundant distance checks for geometrical objects that span multiple cells by first gathering candidates into an unordered set and then filtering them by distance.

1. **Gather Candidates First**:
   - Modified the nested loops to gather candidates instead of directly calculating distances.

2. **Filter Candidates by Distance**:
   - After gathering candidates, a separate loop calculates distances and filters the geometrical objects based on the specified radius and tolerance.

3. **Optimization Details**:
   - Reduced redundant distance calculations by ensuring each geometrical object is processed only once.
   - Improved the efficiency of the algorithm, particularly in scenarios where geometrical objects span multiple cells.

**🆕 Changelog**

- [Refactor GeometricalObjectsBins::SearchInRadius method](https://github.com/KratosMultiphysics/Kratos/commit/64ca24898f3d9654b4caaf237cabf90a03f3e83b)
- [More refactor GeometricalObjectsBins::SearchInRadius method](https://github.com/KratosMultiphysics/Kratos/commit/cd5b11dcd8ee2f138ab52abef41514a694c3993e)

### CODE REVIEW

**Review by roigcarlo — COMMENTED** [review]
👍 for the optimization, but do we really want to delete the SearchInCell function?. It was used as an auxiliar here, but I am not sure if anyone else uses it independently.

**Comment by loumalouomega** [comment]
> 👍 for the optimization, but do we really want to delete the SearchInCell function?. It was used as an auxiliar here, but I am not sure if anyone else uses it independently.

Is still used in other functions

**Comment by roigcarlo** [comment]
What I am saying is that it was public so virtually anyone could, even if we do not

**Comment by loumalouomega** [comment]
> What I am saying is that it was public so virtually anyone could, even if we do not

Okay I wil revert it just in case

**Comment by loumalouomega** [comment]
> > What I am saying is that it was public so virtually anyone could, even if we do not
> 
> Okay I wil revert it just in case

I am thinking is physically impossible to use those methods outside the class. The arguments are super specific for internal use.

File: kratos/spatial_containers/geometrical_objects_bins.cpp [hunk 1]
```diff
@@ -71,30 +71,45 @@ void GeometricalObjectsBins::SearchInRadius(
...
+        auto& r_geometry = p_geometrical_object->GetGeometry();
+        const double distance = r_geometry.CalculateDistance(rPoint, mTolerance);
+        if((Radius + mTolerance) > distance) {
+            rResults.push_back(ResultType(p_geometrical_object));
+            rResults.back().SetDistance(distance);
+        }
     }
+
+    // Shrink the results
+    rResults.shrink_to_fit();
```
Comment (pooyan-dadvand) [inline kratos/spatial_containers/geometrical_objects_bins.cpp]: Do we need to shrink it? This is an slow operation with memory allocation. Usually there are few objects in a cell that are not in the given distance

Comment (loumalouomega) [reply kratos/spatial_containers/geometrical_objects_bins.cpp]: I can simply remove it

**Comment by roigcarlo** [comment]
> I am thinking is physically impossible to use those methods outside the class. The arguments are super specific for internal use.

If you are sure then ok. you have used this more than I did in the last years so 👍

**Review by pooyan-dadvand — APPROVED** [review]
Thanks!

**Pull request merged**